File: javaAPI/src/main/java/com/kotlinjavademo/javaAPI/controllers/KotlinBenefitsController.java
package com.kotlinjavademo.javaAPI.controllers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/kotlin-benefits")
class KotlinBenefitsController {
  @GetMapping("/hey-buddy")
  public String helloWorld() {
    return "Hey Buddy! Love, Kotlin";
  }

  //null is a separate type
  @GetMapping("/nullable")
  public String NullOperators() {
    //here we can still pass back null, but this may not be expected, but it still compiles
    //other times in java we will get a null pointer exception thrown at run time
    boolean toggle = false;
    return toggle? "string" : null;
  }

  // #### Eager Lists ####
  @GetMapping("/lists")
  public List<Integer> lists(){
    List<Integer> list = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      list.add(i * 2);
    }
    List<Integer> filteredList = new ArrayList<>();
    for (int integer : list) {
      if (integer % 4 == 0)
        filteredList.add(integer);
    }
    return list.stream()
        .filter(integer -> integer % 4 ==0)
        .collect(Collectors.toList());
  }

  // the nicere way in java is you can convert to a stream and then back to a list
  @GetMapping("/lists-stream")
  public List<Integer> listToStream(){
    List<Integer> list = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      list.add(i * 2);
    }
    return list.stream()
        .filter(integer -> integer % 4 == 0)
        .collect(Collectors.toList());
  }


  //#### Lazy Streams ####
  @GetMapping("/streams")
  public Stream<Integer> streams() {
    Stream<Integer> stream = Stream.iterate(0,integer -> integer).limit(100);
    return stream.filter(integer -> integer >= 20 && integer <= 40)
                .map(integer -> integer + 10 );
  }

  //#### switch statements ####

  @GetMapping("/switch")
  public String switches(){
    var input = 1;
    switch (input){
      case 0:
        return "switch statements in Java...";
      case 1:
        return "...are limited to a single variable...";
      default:
        return "...and constant value statements only";
    }
  }

}

//#### Default Values ####
// you can set default values with variables in the classes in java
// but then requires it requires a lot of constructors being written and getters and setters
class Person {
  private String title;
  private String firstName = null;
  private String lastName;
  private int age;

  public Person(String title, String firstName, String lastName, int age) {
    this.title = title;
    this.firstName = firstName;
    this.lastName = lastName;
    this.age = age;
  }

  public Person(String title, String lastName, int age) {
    this.title = title;
    this.lastName = lastName;
    this.age = age;
  }

  public Person(String title, String firstName, String lastName) {
    this.title = title;
    this.firstName = firstName;
    this.lastName = lastName;
    this.age = -1;
  }

  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public String getFirstName() {
    return firstName;
  }

  public void setFirstName(String firstName) {
    this.firstName = firstName;
  }

  public String getLastName() {
    return lastName;
  }

  public void setLastName(String lastName) {
    this.lastName = lastName;
  }

  public int getAge() {
    return age;
  }

  public void setAge(int age) {
    this.age = age;
  }
}

File: kotlin/src/main/kotlin/com/kotlinjavademo/kotlin/controllers/KotlinBenefitsController.kt
package com.kotlinjavademo.kotlin.controllers

import com.kotlinjavademo.kotlin.services.JavaService
import org.springframework.web.bind.annotation.*
import java.lang.IllegalArgumentException

@RestController
@RequestMapping("/java-benefits")
class KotlinBenefitsController {
    @GetMapping("/hey-buddy")
    fun helloWorld() : String = "Hey Buddy! Love, Kotlin"

    //null is a separate type
    @GetMapping("/nullable")
    fun nullOperators(): String? {
        // String (non-null String) and String? (nullable String) are two distinct types in Kotlin.
        // This means that the program will not throw a null pointer exception
        // If necessary you can add the operator '!!' after the variable to force a NPE(NullPointerException) check and throw an exception if null

        // if the return type is changed to to just 'String' then the program will not compile
        val toggle = false
        return if (toggle) "string" else null
    }

    // Eager Lists, behave like streams but are executed immediately and the program will wait
    @GetMapping("/lists")
    fun streams() : List<Int>{
        val list = List(5){it * 2}
        return list.filter { it % 4 == 0}
    }

    //Lazy Sequences, the alternative to java streams useful for if you want to add multiple filters or maps
    @GetMapping("/sequences")
    fun sequences(): Sequence<Int>{
        val list = List(100){it}
        return list.asSequence()
                .filter { it in 20..40 }
                .map { it + 10 }
    }

//#### Switch statements ####
    @GetMapping("/switch")
    fun switch(): Any {
        val input:Any = 1
        when(input){
            is String -> return "The input is: $input"
            0 -> return input
            1 -> return 25 + 50
            throw IllegalArgumentException("input doesn't match any cases") -> return "Unreachable code"
        }
    }


    //interesting feature of when is that 'when' can be used without an argument.
    // In such case it acts as a nicer if-else chain: the conditions are Boolean expressions.
    // As always, the first branch that matches is chosen.
    // Given that these are boolean expression, it means that the first condition that results True is chosen.
    @GetMapping("/switch-no-argument")
    fun switchNoArgument(): String{
        val number = 5
        val text = "Hello"

        return when {
            number > 5 -> "Number is greater than 5"
            text.toLowerCase() == "hello" -> "Hi!"
            else -> "you hit the else block"
        }
    }

    // #### Integration with Java ####
    @GetMapping("/java-classes")
    fun javaClasses(): String {
        val javaService = JavaService()
        return javaService.javaMethod()
    }
}

//#### Default Values ####
// default values can be used so that parameters become optional for data classes, this can be use for data classes to avoid lots of constructors
data class Person (
        val title : String,
        val firstName : String? = null,
        val lastName : String,
        val age : Int = -1
)
//In summary, It has a better type system for enhanced safety and compile-time error checking.
// The fact alone that it is null-safe by default should be enough to justify the use of Kotlin over Java.
